src/main/java/fact/gainservice/GainService.java

```diff
@@ -0,0 +1,146 @@
+import fact.io.hdureader.FITS;
+import fact.io.hdureader.OptionalTypesMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import stream.Data;
+import stream.annotations.Parameter;
+import stream.io.CsvStream;
+import stream.io.SourceURL;
+import stream.service.Service;
+
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.net.URL;
+import java.time.Duration;
+import java.time.ZonedDateTime;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * This Service provides gains for each pixel for a given timestamp
+ * It needs a fits file with a BinTable "Gain" which contains the columns "timestamp" and "gain"
+ * timestamp needs to be an ISO-8601 datetime string and "gain" needs to be a length 1440 double array
+ * with the gain values for each pixel.
+ */
+public class GainService implements Service{
+
+    @Parameter(defaultValue = "classpath:/gains_20130825-20170917.fits.gz")
+    public URL gainFile = GainService.class.getResource("/gains_20130825-20170917.fits.gz");
+
+    TreeMap<ZonedDateTime, double[]> gains;
+    private double[] gainsSimulations = null;
+
+    private static final Logger log = LoggerFactory.getLogger(GainService.class);
+
+    /**
+     * Get the closest gain measurement for a given timestamp
+     * @param timestamp
+     * @return gain array for given timestamp
+     */
+    public double[] getGains(ZonedDateTime timestamp) {
+        if (gains == null) {
+            try {
+                loadGains();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+        Map.Entry<ZonedDateTime, double[]> ceiling = gains.ceilingEntry(timestamp);
+        Map.Entry<ZonedDateTime, double[]> floor = gains.floorEntry(timestamp);
+
+        Duration diff;
+        double[] gain;
+
+        if (ceiling == null) {
+            diff = Duration.between(floor.getKey(), timestamp).abs();
+            gain = floor.getValue();
+        } else {
+            if (floor == null) {
+                diff = Duration.between(ceiling.getKey(), timestamp).abs();
+                gain = ceiling.getValue();
+            } else {
+                Duration ceilingDiff = Duration.between(timestamp, ceiling.getKey()).abs();
+                Duration floorDiff = Duration.between(timestamp, floor.getKey()).abs();
+
+                if (ceilingDiff.compareTo(floorDiff) < 0) {
+                    diff = ceilingDiff;
+                    gain = ceiling.getValue();
+                } else {
+                    diff = floorDiff;
+                    gain = floor.getValue();
+                }
+            }
+        }
+
+        if (diff.toDays() >= 5) {
+            log.warn("Time difference to closest gain measurement is more than 5 days");
+        }
+        return gain;
+    }
+
+
+    public double[] getSimulationGains() {
+        if (gainsSimulations == null) {
+            loadGainsSimulations();
+        }
+        return gainsSimulations;
+    }
```

Comment:
Is this getter really needed since it only calls loadGainsSimulations

Reply:
ähm. I think so.

It does not only call `loadGainsSimulation`. It only executes `loadGainsSimulations` if it hasn't already done it. So the getter is necessary to not always read the csv file but only read it once.